packages/babel-plugin-bugfix-v8-static-class-fields-redefine-readonly/src/util.ts

```diff
@@ -0,0 +1,220 @@
+import type { NodePath, Visitor } from "@babel/traverse";
+import { types as t } from "@babel/core";
+
+function isNameOrLength(key: t.Node): boolean {
+  if (t.isIdentifier(key)) {
+    return key.name === "name" || key.name === "length";
+  }
+  if (t.isStringLiteral(key)) {
+    return key.value === "name" || key.value === "length";
+  }
+  return false;
+}
+
+function isStaticFieldWithValue(
+  node: t.Node,
+): node is t.ClassProperty | t.ClassPrivateProperty {
+  return (
+    (t.isClassProperty(node) || t.isClassPrivateProperty(node)) &&
+    node.static &&
+    !!node.value
+  );
+}
+
+const hasReferenceVisitor: Visitor<{ name: string; ref: () => void }> = {
+  ReferencedIdentifier(path, state) {
+    if (path.node.name === state.name) {
+      state.ref();
+      path.stop();
+    }
+  },
+  Scope(path, { name }) {
+    if (path.scope.hasOwnBinding(name)) {
+      path.skip();
+    }
+  },
+};
+
+function isReferenceOrThis(node: t.Node, name?: string) {
+  return t.isThisExpression(node) || (name && t.isIdentifier(node, { name }));
+}
+
+const hasReferenceOrThisVisitor: Visitor<{ name?: string; ref: () => void }> = {
+  "ThisExpression|ReferencedIdentifier"(path, state) {
+    if (isReferenceOrThis(path.node, state.name)) {
+      state.ref();
+      path.stop();
+    }
+  },
+  FunctionParent(path, state) {
```

Comment:
We should not skip ArrowFunctionExpression if we are also looking for ThisExpression.